src/ucp/core/ucp_worker.c

```diff
@@ -2419,6 +2428,87 @@ static void ucp_worker_set_max_am_header(ucp_worker_h worker)
                             ucs_min(max_am_header, UINT32_MAX) : 0ul;
 }
 
+void ucp_worker_track_ep_usage(ucp_request_t *req)
+{
+    static unsigned sample_count_per_round = 100000;
+    ucp_worker_h worker;
+    ucs_time_t now;
+
+    if (ucs_likely((req->send.ep->worker->usage_tracker.iter_count++ %
+                    UCP_WORKER_PROGRESS_TIMER_SKIP_COUNT) != 0)) {
+        return;
+    }
+
+    worker = req->send.ep->worker;
+    now    = ucs_get_time();
+
+    if (ucs_likely((now - worker->usage_tracker.last_round) <
+                   worker->context->config.ext.dynamic_tl_switch_interval)) {
+        return;
+    }
+
+    if (ucs_unlikely(req->flags & UCP_REQUEST_FLAG_USAGE_TRACKED)) {
+        return;
+    }
+
+    req->flags |= UCP_REQUEST_FLAG_USAGE_TRACKED;
+    ucs_usage_tracker_touch_key(worker->usage_tracker.handle, req->send.ep);
+    worker->usage_tracker.sample_count++;
+
+    if (ucs_unlikely((worker->usage_tracker.sample_count %
+                      sample_count_per_round) == 0)) {
+        ucs_usage_tracker_progress(worker->usage_tracker.handle);
+        worker->usage_tracker.last_round = now;
```

Comment:
shouldn't we do it outside the if condition?

Reply:
you want to take 100K samples in each round